04_data_structures/07_roaring_bitmaps/rle.go

```diff
@@ -1,11 +1,232 @@
 package bitmap
 
+import (
+	"encoding/binary"
+	"fmt"
+)
+
+// ==========
+// First idea
+// ==========
+// Each byte contains either compressed or uncompressed data:
+// Bytes starting with 0 are uncompressed bytes, with
+// leading 0 followed by sequence of literal 0s and 1s,
+// 01001011 1000100
+//     ^^^^ ^^^^^^^ - up to 12 bits of uncompressed data
+//  ^^^             - length of uncompressed data, [1; 15]
+// ^                - uncompressed byte flag
+
+// In our scheme we start to make profit only if run is longer than 8,
+// thus we can use bits encoding the length to encode values starting from 8
+// Bytes starting with 1 are compressed bytes:
+// 10100001
+//        ^ - value of the run
+//  ^^^^^^  - length of the run, starting from 2^3 to to 2^9 - 1 = 511
+// ^        - compressed byte flag
+
+// Disadvantages:
+// annoying to work with individual bits
+//
+
+// ===========
+// Second idea
+// ===========
+// Let's work with whole bytes instead:
+
+// 0_______
+//  ^^^^^^^ - length of followed uncompressed bytes[1;127]
+// ^        - uncompressed flag
+
+// 1_______
+//   ^^^^^^ - length of the run in bytes [1;63]
+//  ^       - value, of the run, 0 or 1
+// ^        - compressed flag
+
+type byteType uint
+
+const (
+	compressed0 byteType = iota
+	compressed1
+	uncompressed
+)
+const maxUncompressedRunLength = 127 // 7 bits
+const maxCompressedRunLength = 63    // 6 bits
+
+func getByteType(b byte) byteType {
+	if b == 0x00 {
+		return compressed0
+	}
+	if b == 0xff {
+		return compressed1
+	}
+	return uncompressed
+}
+
+func genHeaderByte(st byteType, runLength uint8) byte {
+	if runLength == 0 {
+		panic("BUG! run length must be non-zero to produce valid encoding")
+	}
+	if st == uncompressed {
+		if runLength > maxUncompressedRunLength {
+			// TODO abozhenko for robot-dreams:
+			// If I know that runLength for uncompressed must be in range [1;127]
+			// and length for compressed byte must be in range [1;63],
+			// how do I enforce these rules at compile time?
+			// In oCaml that can be solved with custom type
+			panic(fmt.Sprintf("BUG! run length(%v) of uncompressed bytes must be encoded with just 6 bits",
+				runLength))
+		}
+		return byte(runLength)
+	}
+	if runLength > maxCompressedRunLength {
+		panic(fmt.Sprintf("BUG! run length(%v) of compressed bytes must be encoded with just 5 bits",
+			runLength))
+	}
+	// TODO abozhenko for robot-dreams
```

Comment:
@robot-dreams

Reply:
Again, unfortunately Go isn't as nice as OCaml at compile time :( probably you would just do a `switch` statement here and have the `default` case return an error, which would only catch the issue at runtime.